python/ml4ir/base/io/file_io.py

```diff
@@ -152,17 +146,22 @@ def write_df(df, outfile: str = None, sep: str = ",", index: bool = True, log=No
 
     Returns:
         dataframe in csv form if outfile is None
+
+    NOTE: Does not support spark write
     """
-    output = df.to_csv(
-        sep=sep, index=index, quotechar='"', escapechar="\\", quoting=csv.QUOTE_NONNUMERIC
-    )
-    output = output.replace("\\", "\\\\")
+    if outfile and outfile.startswith(HDFS_PREFIX):
+        raise NotImplementedError
+    else:
+        output = df.to_csv(
+            sep=sep, index=index, quotechar='"', escapechar="\\", quoting=csv.QUOTE_NONNUMERIC
+        )
+        output = output.replace("\\", "\\\\")
 
-    if outfile:
-        fp = open(outfile, "w")
-        fp.write(output)
-        fp.close()
-    return output
+        if outfile:
+            fp = open(outfile, "w")
```

Comment:
Here using `with open()...` is a more pythonic way. Why not directly write with pd.to_csv? to replace?

Reply:
After spending a lot of time with our ranking data, this combination of `read_df` and `write_df` were the only ones that were compatible with each other and the upstream spark jobs. If you notice, I need to replace `\\` with `\\\\` on the next line for compatibility with re-reading. I can add a `FIXME` here for the future though.

Reply:
(Again, this method was written a year ago for some of our other projects. Might not be necessary to do all this hacky read write for ml4ir)